### ISSUE TITLE
ERR_PNPM_TARBALL_EXTRACT error with mv inside content-addressable store

### ISSUE THREAD

**Issue by VladimirNT** [pnpm/pnpm#7676]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Reproduction steps

The issue is flaky and happens randomly for different packages for us after switching from `pnpm@7` to `pnpm@8.14.1`.
The interesting part is that according to the logs some file in CAS is to be renamed to some reduced length name, which looks a bit weird.
 
We are installing the dependencies using the following keys:

pnpm command example:
```
NODE_RES/node PNPM_RES/node_modules/pnpm/dist/pnpm.cjs install --frozen-lockfile --ignore-pnpmfile --ignore-scripts --no-verify-store-integrity --offline --package-import-method hardlink --store-dir benchmarks/targets/frontend/services/jobs/node_modules/.pnpm/store --strict-peer-dependencies --virtual-store-dir benchmarks/targets/frontend/services/jobs/node_modules/.pnpm/virtual-store --registry http://OUR_INTERNAL_REGISTRY --stream --reporter append-only --no-color
```

### Describe the Bug

Error message example:
```
ERR_PNPM_TARBALL_EXTRACT  Failed to unpack the tarball from "contrib/typescript/available-typed-arrays/-/available-typed-arrays-1.0.5.tgz": Error: ENOENT: no such file or directory, rename 'benchmarks/targets/frontend/services/jobs/node_modules/.pnpm/store/v3/files/54/b84472aba9dc81d7b5924fb74ed962803d24d463cb58e153f354e35630e04f2613279aff3fba6f0e612f796108ed3da638bd134047d90dda0d775cde2f7306263816' -> 'benchmarks/targets/frontend/services/jobs/node_modules/.pnpm/store/v3/files/54/b84472aba9dc81d7b5924fb74ed962803d24d463cb58e153f354e35630e04f2613279aff3fba6f0e612f796108ed3da638bd134047d90dda0d775cde2f7306'
```

### Expected Behavior

Successful modules installation

### Which Node.js version are you using?

18.18.2

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Ubuntu 18.04

**Comment by zkochan** [comment]
What packages were you installing to get the error?

**Comment by VladimirNT** [comment]
> What packages were you installing to get the error?

Hello,

thanks for looking into this.

The projects we are running `pnpm i` onto are our internal ones, having different size (from ~10 to ~5000 dependencies), as I mentioned above it is quite random.

We have it appearing for different packages and we can't see any exact pattern for now. 
These are the packages we noticed the issue with lately:
- https://www.npmjs.com/package/mdast-util-from-markdown/v/0.8.5
- https://www.npmjs.com/package/@babel/plugin-syntax-optional-catch-binding/v/7.8.3
- https://www.npmjs.com/package/@aws-sdk/is-array-buffer/v/3.55.0
- https://www.npmjs.com/package/available-typed-arrays/v/1.0.5

**Comment by zaverden** [comment]
I think I found where is a problem, though I don't know what is a low-level cause of it. I've looked into the problem with `pnpm@9.6.0`.

Error comes from [optimisticRenameOverwrite](https://github.com/pnpm/pnpm/blob/eed2cf95df2e3c553b171e609af87a327d11a1d8/store/cafs/src/writeBufferToCafs.ts#L58). It gets ENOENT and also fs.existsSync(fileDest) says target file does not exists.

I really don't know why this can happen but it looks like race condition. I was not able to reproduce it locally, but it happes pretty often on our build farm. It may be some specificities of hard drives on some servers.

I have patched function [pathTemp](https://github.com/pnpm/pnpm/blob/main/store/cafs/src/writeBufferToCafs.ts#L82). I have added thread id to the temp filename:
```
import { threadId } from 'worker_threads'
 
export function pathTemp (file: string): string {
  const basename = removeSuffix(path.basename(file))
  return path.join(path.dirname(file), `${basename}${process.pid}-${threadId}`)
}
```

I have run build on our 300+ independent projects for several times, and never saw the issue again 

I'll send a PR with the patch

**Comment by VladimirNT** [comment]
@zkochan, please take a look into the solution provided by @zaverden above.
I does look like there were many issues reported with the same root cause and all those can be fixed by avoiding race on tempPath writing.

**Comment by VladimirNT** [comment]
@zkochan I've prepared a fix for the issue and confirmed it with running install for many projects together with 128 workers - it definitely fails without the fix and succeeds with it.
Unfortunately I have no access to the pnpm repo to push the change. I could do that if someone would grant the permissions.
Otherwise one can find the diff here and raise the fixup:
```
Author: VladimirNT
Date:   Thu Oct 24 22:17:34 2024 +0300

    fix(cafs): avoid race on extracting files from tarballs
    
    On running addFilesFromTarball with many workers in the pool there were sometimes
    race conditions on moving files having same content.
    It resulted in ERR_PNPM_ENOENT and ERR_PNPM_TARBALL_EXTRACT exceptions during install.
    The issue was introduced in v8.7.0 and generated many defect reports.
    
    Closes #7676, possibly #7607, #7697 and many others

diff --git a/store/cafs/src/writeBufferToCafs.ts b/store/cafs/src/writeBufferToCafs.ts
index c04f6ead9..357b4eb1b 100644
--- a/store/cafs/src/writeBufferToCafs.ts
+++ b/store/cafs/src/writeBufferToCafs.ts
@@ -1,5 +1,6 @@
 import fs from 'fs'
 import path from 'path'
+import workerThreads from 'worker_threads'
 import util from 'util'
 import renameOverwrite from 'rename-overwrite'
 import type ssri from 'ssri'
@@ -81,7 +82,7 @@ export function optimisticRenameOverwrite (temp: string, fileDest: string): void
  */
 export function pathTemp (file: string): string {
   const basename = removeSuffix(path.basename(file))
-  return path.join(path.dirname(file), `${basename}${process.pid}`)
+  return path.join(path.dirname(file), `${basename}${process.pid}-${workerThreads.threadId}`)
 }
 
 function removeSuffix (filePath: string): string {
```
